packages/core/src/browser/decoration-style.ts

```diff
@@ -0,0 +1,61 @@
+/********************************************************************************
+ * Copyright (C) 2021 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+export namespace DecorationStyle {
+
+    export function createStyleSheet(container: HTMLElement = document.getElementsByTagName('head')[0]): CSSStyleSheet | undefined {
+        if (!container) {
+            return undefined;
+        }
+        const style = document.createElement('style');
+        style.id = 'editorDecorationsStyle';
+        style.type = 'text/css';
+        style.media = 'screen';
+        style.appendChild(document.createTextNode('')); // trick for webkit
+        container.appendChild(style);
+        return <CSSStyleSheet>style.sheet;
+    }
+
+    export function createRule(selector: string, styleProvider: (style: CSSStyleDeclaration) => void,
+        styleSheet: CSSStyleSheet | undefined): void {
+        if (!styleSheet) {
+            return;
+        }
+
+        const index = styleSheet.insertRule(selector + '{}', 0);
+        const rules = styleSheet.cssRules || styleSheet.rules;
+        const rule = rules[index];
+        if (rule && rule.type === CSSRule.STYLE_RULE) {
+            const styleRule = rule as CSSStyleRule;
+            styleProvider(styleRule.style);
+        }
+    }
```

Comment:
A few comments on this method:
 - If you insert rules at index 0, they have the lowest precedence, so adding tow rules for the same selector and rule would result in the earlier rule overriding the later, which probably isn't intended. There are two ways out:
   - figure out the next index and insert at the end of the list.
   - search the list for rules with the same selector and work with the existing rule. This would simplify the `deleteRule` logic a bit, since you would know that (unless the caller did something weird with their style sheet), there would be only one rule for a given selector, and you could bail if/when you found it.
 - The `styleProvider` callback is weird. I would probably just return `CSSStyleRule` and let the caller do what they want with it rather than having them pass in a callback.

Reply:
Thank you @colin-grant-work for the very detailed explanation. I updated the code for `createRule` (renamed to `createStyleRule`) and `deleteRule` (renamed to `deleteStyleRule` to avoid confusion with the other `deleteRule` that exists in the code) so that there is only one rule for a given selector.

I also updated `editor-decoration-style.ts` and removed the `styleProvider` callback based on your comment.